vllm/distributed/parallel_state.py

```diff
@@ -133,25 +144,32 @@ def initialize_model_parallel(
     rank = torch.distributed.get_rank()
 
     # Build the tensor model-parallel groups.
-    global _TP_DEVICE_GROUP, _TP_CPU_GROUP
+    global _TP_DEVICE_GROUP, _TP_CPU_GROUP, _TP_PYNCCL_COMMUNICATOR
     assert _TP_DEVICE_GROUP is None, (
         "tensor model parallel group is already initialized")
     for i in range(num_tensor_model_parallel_groups):
-        ranks = range(i * tensor_model_parallel_size,
-                      (i + 1) * tensor_model_parallel_size)
+        ranks = list(
+            range(i * tensor_model_parallel_size,
+                  (i + 1) * tensor_model_parallel_size))
         group = torch.distributed.new_group(ranks, backend=backend)
         cpu_group = torch.distributed.new_group(ranks, backend="gloo")
         if rank in ranks:
             _TP_DEVICE_GROUP = group
             _TP_CPU_GROUP = cpu_group
 
+    from vllm.distributed.device_communicators.pynccl import NCCLCommunicator
```

Comment:
Again, why do we need this lazy import?

Reply:
lazy import is required here to avoid circular import. `vllm.distributed.device_communicators.pynccl` will try to import `vllm/distributed/parallel_state.py`.

Reply:
If that's the case, I think it means it's a bad design tbh. We should use lazy import only to avoid the unnecessary imports, but to avoid circular imports. Otherwise, the code will be too complicated.